be/src/storage/lake/segment_writer_finalize_executor.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file is based on code available under the Apache license here:
+//   https://github.com/apache/incubator-doris/blob/master/be/src/olap/memtable_flush_executor.h
+
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <vector>
+
+#include "common/status.h"
+#include "util/threadpool.h"
+
+namespace starrocks::lake {
+
+class VerticalGeneralTabletWriter;
+
+// the statistic of a certain finalize handler.
+// use atomic because it may be updated by multi threads
+struct SegmentWriterFinalizeStatistic {
+    int64_t finalize_time_ns = 0;
+    int64_t finalize_count = 0;
+    std::atomic<int64_t> queueing_segment_writer_num = 0;
+};
+
+std::ostream& operator<<(std::ostream& os, const SegmentWriterFinalizeStatistic& stat);
+
+class SegmentWriterFinalizeToken {
+public:
+    explicit SegmentWriterFinalizeToken(std::unique_ptr<ThreadPoolToken> thread_pool_token);
+
+    ThreadPoolToken* get_thread_pool_token() { return _thread_pool_token.get(); }
+
+    void cancel();
+
+    void wait();
+
+    const SegmentWriterFinalizeStatistic& get_stats() const { return _stats; }
+
+private:
+    friend class VerticalGeneralTabletWriter;
+
+    std::unique_ptr<ThreadPoolToken> _thread_pool_token;
+    SegmentWriterFinalizeStatistic _stats;
```

Comment:
looks like this WriterFinalizeToken wraps ThreadPoolToken with additional _stats, but these _stats is not used anywhere, is there subsequent PR coming for it?

Reply:
have updated to reuse memtable's flush pool